---
title: "Post DEG analysis"
output: html_document
---

This script will analyze the results from the differential expression analysis

## Preliminaries

```{r libraries, message=FALSE}
library(GenomicAlignments)
library(gage)
library(pathview)
library(gageData)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(dplyr)
```

```{r databases, message=FALSE}
setwd('..')
path <- getwd()
setwd(paste(path, "/Results/DEGenes", sep = ""))
load("resmatSig_Adipose - Subcutaneous_.RData")
```

## Descriptive stats

```{r summaries}
summary(ResSig)
```

## Gene Set Enrichment Analysis (GSEA)

```{r testing}
geneids <- gsub("\\..*","",row.names(ResSig))

ResSig$symbol  <- mapIds(org.Hs.eg.db,
                     keys=geneids,
                     column="SYMBOL",
                     keytype="ENSEMBL",
                     multiVals="first")
ResSig$entrez  <- mapIds(org.Hs.eg.db,
                     keys=geneids,
                     column="ENTREZID",
                     keytype="ENSEMBL",
                     multiVals="first")
ResSig$name  <- mapIds(org.Hs.eg.db,
                     keys=geneids,
                     column="GENENAME",
                     keytype="ENSEMBL",
                     multiVals="first")
```

```{r}
data(kegg.sets.hs)
data(sigmet.idx.hs)
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]
head(kegg.sets.hs, 3)

foldchanges <- ResSig$log2FoldChange
names(foldchanges) <- ResSig$entrez
head(foldchanges)

keggres = gage(foldchanges, gsets=kegg.sets.hs, same.dir=TRUE)
lapply(keggres, head)

keggrespathways = data.frame(id=rownames(keggres$greater), keggres$greater) %>% 
  tbl_df() %>% 
  filter(row_number()<=5) %>% 
  .$id %>% 
  as.character()
keggrespathways
keggresids = substr(keggrespathways, start=1, stop=8)
keggresids
plot_pathway = function(pid) pathview(gene.data=foldchanges, pathway.id=pid, species="hsa", new.signature=FALSE)
tmp = sapply(keggresids, function(pid) pathview(gene.data=foldchanges, pathway.id=pid, species="hsa"))
```

