---
title: "Differentially expressed genes per tissue"
output: html_document
---

In this code we will estimate differential expression for each tissue and then export the final list of genes.
Briefly, we will remove genes with low count for each tissue, and run DESeq2 pipeline to estimate differential expression.

## Preliminaries

Loading libraries

```{r libraries, message=FALSE}
library(DESeq2)
library(edgeR)
```

Loading databases

```{r databases, message=FALSE, warning=FALSE}
setwd('..')
path <- getwd()
setwd(paste(path, "/Results/GTEx", sep = ""))
samples <- read.csv("SubjSample.csv")
counts  <- read.csv(gzfile("GeneReads.csv.gz"), row.names=1)
head(samples)
counts[1:4,1:4]
```

## Estimating DE genes

This is the core of the script.
For each tissue we will run the following pipeline:
- First, we will transform the gene counts into counts per million (cpm) values, and remove those with values lower than the n of the minimum group size
- Second, and following the DESeq2 pipeline, we will compute the differential expression analysis, using Subject, Age, and Sex as covariates
- Finally, a table with the results will be exported as .RData file

```{r DEG by tissue, message=FALSE}
setwd(paste(path, "/Results/DEGenes", sep = ""))

SamplesByTissue <- split(samples, samples$SMTSD)

for(tis in 1:(length(SamplesByTissue)-44)){
  GeneCounts <- counts[,colnames(counts) %in% SamplesByTissue[[tis]]$SAMPID] #Select reads of the selected tissue
  keep       <- rowSums(cpm(GeneCounts)>1) >= min(summary(as.factor(SamplesByTissue[[tis]]$GENDER))) #Remove genes with low counts, where the second 
                                                                                                     #factor is the n of the group with less subjects
  symbols    <- as.character(counts[keep,1])
  GeneCounts <- GeneCounts[keep,] #Removing low counts from GeneCounts file
  insample   <- SamplesByTissue[[tis]]$SAMPID %in% colnames(GeneCounts)
  cond       <- SamplesByTissue[[tis]][insample, 2:4] #Create covariate conditions (Gender and age)
  rownames(cond) <- SamplesByTissue[[tis]][insample, 1] #Adding sample ID as rowname 
  cond$GENDER    <- as.factor(cond$GENDER) #making gender a factor
  #Order cond and GeneCounts to match each other
  cond       <- cond[order(rownames(cond)), ]
  GeneCounts <- GeneCounts[, order(colnames(GeneCounts))]
  print(as.character(SamplesByTissue[[tis]]$SMTSD[1])) #Printing the tissue 
  #DESeq analysis
  mat <- DESeqDataSetFromMatrix(countData = GeneCounts,  #Creating deseq database with counts and covariates (cond)
                                colData = cond,
                                design  = ~ GENDER + AGE)
  mat    <- DESeq(mat) #DE estimation
  resmat <- results(mat, contrast = c("GENDER", "1", "2")) #Export the results, with contrast between males and females. 
  #P adjustment can be changed, this case is BH method
  resmat$symbol <- symbols
  print(table(resmat$padj < 0.05)) #Number of differentially expressed genes
  ResSig <- subset(resmat, padj < 0.05) #Subset results to those differentially expressed only
  save(ResSig, file = paste("resmatSig", as.character(SamplesByTissue[[tis]]$SMTSD[1]), ".RData", sep = "_")) #Saving DEG as .RData
}
```

